### PR TITLE
[Product][Attribute][Association] Product-related bundles clean up

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | yes
| Related tickets |
| License         | MIT

- [x] Association
- [x] AssociationBundle
- [x] Attribute
- [x] AttributeBundle
- [x] Product
- [x] ProductBundle

### CODE REVIEW

File: src/Sylius/Component/Association/Model/Association.php [hunk 1]
```diff
@@ -39,7 +38,7 @@ class Association implements AssociationInterface
     protected $owner;
 
     /**
-     * @var Collection<AssociableInterface>
+     * @var AssociableInterface[]
```
Comment (tuka217) [inline src/Sylius/Component/Association/Model/Association.php]: Shouldn't be Collection|AssociableInterface[] ?

File: src/Sylius/Component/Association/Model/AssociationInterface.php [hunk 2]
```diff
@@ -36,27 +35,27 @@ public function setType(AssociationTypeInterface $type);
...
 
     /**
-     * @param AssociableInterface
+     * @param AssociableInterface|null $owner
      */
     public function setOwner(AssociableInterface $owner = null);
 
     /**
-     * @return Collection
+     * @return AssociableInterface[]
```
Comment (tuka217) [inline src/Sylius/Component/Association/Model/AssociationInterface.php]: Same here

File: src/Sylius/Bundle/AttributeBundle/DependencyInjection/Compiler/RegisterAttributeFactoryPass.php [hunk 3]
```diff
@@ -32,7 +32,7 @@ public function process(ContainerBuilder $container)
 
         $registry = $container->getDefinition('sylius.registry.attribute_type');
 
-        foreach ($container->getParameter('sylius.attribute.subjects') as $subject => $configuration) {
+        foreach (array_keys($container->getParameter('sylius.attribute.subjects')) as $subject) {
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/AttributeBundle/DependencyInjection/Compiler/RegisterAttributeFactoryPass.php]: Do we need `array_keys` here?

File: src/Sylius/Bundle/ProductBundle/Doctrine/ORM/ProductVariantRepository.php [hunk 4]
```diff
@@ -32,7 +32,8 @@ public function findByName($name)
      */
     public function createQueryBuilderWithProduct($productId)
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ProductBundle/Doctrine/ORM/ProductVariantRepository.php]: Is this method used anywhere?

File: src/Sylius/Bundle/ProductBundle/Validator/ProductVariantCombinationValidator.php [hunk 5]
```diff
@@ -50,16 +50,12 @@ public function validate($value, Constraint $constraint)
     private function matches(ProductVariantInterface $variant, ProductInterface $variable)
     {
         foreach ($variable->getVariants() as $existingVariant) {
-            if ($variant === $existingVariant) {
+            if ($variant === $existingVariant || !$variant->getOptionValues()->count()) {
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ProductBundle/Validator/ProductVariantCombinationValidator.php]: `0 === $variant->getOptionValues()->count()`

**Comment by michalmarcinkowski** [comment]
Good job Mateusz! 👍

File: src/Sylius/Component/Association/Model/Association.php (hunk 1 above)
Comment (Zales0123) [inline src/Sylius/Component/Association/Model/Association.php]: I'm not sure, as it implied that there can be collection of ``AssociableInterface`` objects **or** some

File: src/Sylius/Bundle/ProductBundle/Doctrine/ORM/ProductVariantRepository.php (hunk 4 above)
Comment (Zales0123) [inline src/Sylius/Bundle/ProductBundle/Doctrine/ORM/ProductVariantRepository.php]: Yes, in ``sylius_admin_product_variant`` grid.

File: src/Sylius/Bundle/ProductBundle/Form/Type/ProductType.php [hunk 6]
```diff
@@ -30,7 +30,7 @@ class ProductType extends AbstractResourceType
     private $variantResolver;
 
     /**
-     * @param string $dataClass FQCN
+     * @param string $dataClass
```
Comment (Zales0123) [inline src/Sylius/Bundle/ProductBundle/Form/Type/ProductType.php]: Then we should add it in every form type with defined/overwritten constructor :) In my opinion it's quite obvious and most of all we should be consistent with this in all form types.

Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ProductBundle/Form/Type/ProductType.php]: Agree that it should be obvious for everyone, no need to bring it back.

**Pull request merged**